lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntime.cpp

```diff
@@ -476,51 +480,46 @@ void SwiftLanguageRuntimeImpl::SetupReflection() {
-      m_initialized_reflection_ctx = false;
-      return;
-    }
 
-    bool objc_interop = (bool)findRuntime(m_process, RuntimeKind::ObjC);
-    const char *objc_interop_msg =
-        objc_interop ? "with Objective-C interopability" : "Swift only";
+  if (m_initialized_reflection_ctx)
+    return;
 
-    auto &triple = exe_module->GetArchitecture().GetTriple();
-    if (triple.isArch64Bit()) {
-      LLDB_LOGF(GetLog(LLDBLog::Types),
-                "Initializing a 64-bit reflection context (%s) for \"%s\"",
-                triple.str().c_str(), objc_interop_msg);
-      m_reflection_ctx = ReflectionContextInterface::CreateReflectionContext64(
-          this->GetMemoryReader(), objc_interop, GetSwiftMetadataCache());
-    } else if (triple.isArch32Bit()) {
-      LLDB_LOGF(GetLog(LLDBLog::Types),
-                "Initializing a 32-bit reflection context (%s) for \"%s\"",
-                triple.str().c_str(), objc_interop_msg);
-      m_reflection_ctx = ReflectionContextInterface::CreateReflectionContext32(
-          this->GetMemoryReader(), objc_interop, GetSwiftMetadataCache());
-    } else {
-      LLDB_LOGF(GetLog(LLDBLog::Types),
-                "Could not initialize reflection context for \"%s\"",
-                triple.str().c_str());
-    }
+  if (!exe_module) {
+    LLDB_LOGF(GetLog(LLDBLog::Types), "%s: Failed to get executable module",
+              LLVM_PRETTY_FUNCTION);
+    m_initialized_reflection_ctx = false;
+    return;
+  }
 
-    modules_snapshot = std::move(m_modules_to_add);
-    m_modules_to_add.Clear();
-    m_initialized_reflection_ctx = true;
+  bool objc_interop = (bool)findRuntime(m_process, RuntimeKind::ObjC);
+  const char *objc_interop_msg =
+      objc_interop ? "with Objective-C interopability" : "Swift only";
+
+  auto &triple = exe_module->GetArchitecture().GetTriple();
+  if (triple.isArch64Bit()) {
+    LLDB_LOGF(GetLog(LLDBLog::Types),
+              "Initializing a 64-bit reflection context (%s) for \"%s\"",
+              triple.str().c_str(), objc_interop_msg);
+    m_reflection_ctx = ReflectionContextInterface::CreateReflectionContext64(
+        this->GetMemoryReader(), objc_interop, GetSwiftMetadataCache());
+  } else if (triple.isArch32Bit()) {
+    LLDB_LOGF(GetLog(LLDBLog::Types),
+              "Initializing a 32-bit reflection context (%s) for \"%s\"",
+              triple.str().c_str(), objc_interop_msg);
+    m_reflection_ctx = ReflectionContextInterface::CreateReflectionContext32(
+        this->GetMemoryReader(), objc_interop, GetSwiftMetadataCache());
+  } else {
+    LLDB_LOGF(GetLog(LLDBLog::Types),
+              "Could not initialize reflection context for \"%s\"",
+              triple.str().c_str());
   }
+  // A copy of the modules that should be added, to prevent mutating the list
+  // while iterating over it.
+  ModuleList modules_snapshot = std::move(m_modules_to_add);
+  m_modules_to_add.Clear();
+  m_initialized_reflection_ctx = true;
 
   Progress progress(
```

Comment:
Can you add a comment here explaining that AddModuleToReflectionContext will call into SetupReflection again, but it will succeed because we already m_initialized_reflection_ctx. However, other threads are blocked from running it because of the mutex?